### PR TITLE
Refactor training/eval

### PR DESCRIPTION
**Pull request by ealt**

### CODE REVIEW

File: simplexity/configs/config.py [hunk 1]
```diff
@@ -1,25 +1,90 @@
...
     training: TrainingConfig
-    validation: ValidationConfig
+    validation: ValidationConfig | None
 
     seed: int
     experiment_name: str
     run_name: str
+
+
+def validation_required(cfg: Config) -> bool:
```
Comment (hrbigelow) [inline simplexity/configs/config.py:35]: It can be nice to use `__post_init__` method of the Config object instead of these stand-alone functions.

File: simplexity/configs/predictive_model/config.py [hunk 2]
```diff
@@ -1,19 +1,22 @@
...
 
+TARGETS = Literal["simplexity.predictive_models.gru_rnn.build_gru_rnn"]
+
 
 @dataclass
 class ModelInstanceConfig:
     """Configuration for the model instance."""
 
-    _target_: Literal["simplexity.predictive_models.rnn.build_rnn"]
+    _target_: TARGETS
```
Comment (hrbigelow) [inline simplexity/configs/predictive_model/config.py:11]: I thought that the _target_ field is only something you put in the yaml hydra config file, isn't it?

File: simplexity/evaluation/evaluate_model.py [hunk 3]
```diff
@@ -1,36 +1,31 @@
...
-    return {"loss": mean_batch_loss, "accuracy": mean_batch_accuracy}
+def evaluation_step(
+    model: PredictiveModel, inputs: jax.Array, labels: jax.Array, metric_keys: Iterable[str] = ("loss", "accuracy")
+) -> dict[str, jax.Array]:
+    """Cross entropy loss for a penzai model."""
+    logits = model(inputs)
+    metrics = {}
+    for metric_key in metric_keys:
+        metric_values = METRIC_FUNCTIONS[metric_key](logits, labels)
+        metrics[metric_key] = jnp.mean(metric_values)
```
Comment (hrbigelow) [inline simplexity/evaluation/evaluate_model.py:23]: This is also too strong of an assumption here - some of the metrics will want to take the mean over batch but not over context.

File: simplexity/predictive_models/gru_rnn.py [hunk 4]
```diff
@@ -6,13 +6,32 @@
...
+        self.embedding = eqx.nn.Embedding(vocab_size, embedding_size, key=key)
+
+    def __call__(self, xs: jax.Array, **kwargs) -> jax.Array:
+        """Forward pass of the linear model."""
+
+        def process_element(_, x):
+            out = self.embedding(x)
+            return None, out
+
+        _, outs = jax.lax.scan(process_element, None, xs)
```
Comment (hrbigelow) [inline simplexity/predictive_models/gru_rnn.py:24]: You want vmap instead of scan here since there is no carry.  Using scan will convert this into a sequential computation.

File: simplexity/run.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+from simplexity.generative_processes.generative_process import GenerativeProcess
+from simplexity.logging.logger import Logger
+from simplexity.persistence.model_persister import ModelPersister
+from simplexity.predictive_models.predictive_model import PredictiveModel
+from simplexity.training.train_model import train
+from simplexity.utils.hydra import typed_instantiate
+
+
+@hydra.main(config_path="configs", config_name="train_model.yaml", version_base="1.2")
+def train_model(cfg: Config) -> float:
```
Comment (hrbigelow) [inline simplexity/run.py:14]: Should be `train_model(cfg: DictConfig)` - and no need to put that assert in there I think.

File: simplexity/run.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    else:
+        validation_data_generator = None
+        validation_bos_token = None
+        validation_eos_token = None
+
+    vocab_size = training_data_generator.vocab_size
+    model = typed_instantiate(cfg.predictive_model.instance, PredictiveModel, vocab_size=vocab_size)
+
+    if cfg.persistence:
+        with typed_instantiate(cfg.persistence.instance, ModelPersister) as persister:
```
Comment (hrbigelow) [inline simplexity/run.py:41]: This feels quite verbose - can't you conditionally instantiate the persister or None, and then not just call `train` once.

File: simplexity/training/equinox_trainer.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+
+class EquinoxTrainer(eqx.Module):
+    """A trainer for a predictive model."""
+
+    root_rng: chex.PRNGKey
+    model: pz.StateVariable[PredictiveModel]
+    state: pz.StateVariable[InternalTrainerState]
+    optimizer_def: optax.GradientTransformation = dataclasses.field(metadata={"pytree_node": False})
+    loss_fn: LossFunction = dataclasses.field(metadata={"pytree_node": False})
```
Comment (hrbigelow) [inline simplexity/training/equinox_trainer.py:74]: At some point can you also add option to have a learning rate schedule funciton?

File: simplexity/training/train_model.py [hunk 8]
```diff
@@ -51,18 +49,40 @@ def train(
...
-) -> tuple[PenzaiModel, float]:
+) -> tuple[PredictiveModel, float]:
     """Train a predictive model on a generative process."""
     key = jax.random.PRNGKey(training_cfg.seed)
 
+    def get_model():
+        return model
+
     key, trainer_key = jax.random.split(key)
     optimizer = typed_instantiate(training_cfg.optimizer.instance, optax.GradientTransformation)
```
Comment (hrbigelow) [inline simplexity/training/train_model.py:60]: I think it's a bit messy to be instantiating sub-configs this late in the program.  Preferably you would just instantiate a global Config object at the very beginning of the program from a DictConfig using `hydra.utils.instantiate`.

File: simplexity/training/train_model.py [hunk 9]
```diff
@@ -81,20 +101,25 @@ def train(
...
+                    data_generator=validation_data_generator,
                     bos_token=validation_bos_token,
                     eos_token=validation_eos_token,
                 )
                 validation_metrics = {f"validation/{k}": v for k, v in validation_metrics.items()}
                 logger.log_metrics(step, validation_metrics)
-        if persister and step % training_cfg.checkpoint_every == 0:
-            persister.save_weights(model, step)
+        if persister and training_cfg.checkpoint_every and step % training_cfg.checkpoint_every == 0:
+            persister.save_weights(get_model(), step)
```
Comment (hrbigelow) [inline simplexity/training/train_model.py:122]: As a TODO - if you intend to resume training from this checkpoint, you should also checkpoint the optimizer state.

**Review by hrbigelow — APPROVED** [review]
Please see comments but feel free to merge.  I didn't see any obvious bugs.

File: simplexity/configs/config.py (hunk 1 above)
Comment (ealt) [reply simplexity/configs/config.py:35]: I don't believe Hydra actually uses these user defined dataclasses to instantiate the configs, they just exist to provide type hints for IDE support and static analysis

File: simplexity/configs/predictive_model/config.py (hunk 2 above)
Comment (ealt) [reply simplexity/configs/predictive_model/config.py:11]: It is part of the hydra file, and these dataclasses are meant to be 1:1 mappings to the yaml configs

File: simplexity/configs/config.py [hunk 10]
```diff
@@ -1,25 +1,92 @@
...
 from simplexity.configs.persistence.config import Config as PersistenceConfig
 from simplexity.configs.predictive_model.config import Config as ModelConfig
+from simplexity.configs.predictive_model.config import validate_config as validate_model_config
 from simplexity.configs.training.config import Config as TrainingConfig
+from simplexity.configs.training.config import validate_config as validate_training_config
 
 
 @dataclass
-class Config:
+class Config(DictConfig):
```
Comment (hrbigelow) [inline simplexity/configs/config.py:21]: This is not necessary.  All of your application's user-defined dataclasses should have no knowledge of hydra at all.  hydra yaml files refer to them in their `_target_` field, and that's the only link.  Then, in the application you just write:

opts: MyCustomDataclass = hydra.utils.instantiate(cfg: DictConfig)

File: simplexity/configs/config.py (hunk 1 above)
Comment (hrbigelow) [reply simplexity/configs/config.py:37]: Yes - hydra.utils.instantiate will recursively call your dataclass constructor and constructors of members.

**Review by hrbigelow — DISMISSED** [review]
I'm approving but please see comment - you don't need to inherit from DictConfig - just pretend hydra doesn't exist.

File: simplexity/training/train_model.py (hunk 8 above)
Comment (ealt) [reply simplexity/training/train_model.py:60]: I generally agree, but I instantiate it here because it is defined as part of the training config that needs to be passed into this function
